### PR TITLE
AoC 2019 Day 1 Part 1

### PR DESCRIPTION
**Pull request by P-on-D**
I've been practicing! 😄 This simple solution shows off so much of what I find attractive about D.

### CODE REVIEW

File: 2019/d01p1FuelCounterUpper.d [hunk 1]
```diff
@@ -0,0 +1,7 @@
+unittest {
```
Comment (P-on-D-Reviewer) [inline 2019/d01p1FuelCounterUpper.d]: You say you've been practicing. Demonstrate this with a show and tell. Help the audience understand why you find this D so attractive.

**Review by P-on-D-Reviewer — CHANGES_REQUESTED** [review]

File: 2019/d01p1FuelCounterUpper.d (hunk 1 above)
Comment (P-on-D-Reviewer) [reply 2019/d01p1FuelCounterUpper.d]: I submit that any programmer can read this and understand what is going on. I further submit that this code _as a solution to the AoC problem_ can be easily explained to a non-programmer.

This line alone says so much! Everything enclosed in the braces is intended to prove something meaningful. This has to be my favourite feature of D so far. I like to "sketch" in code so having a code-based canvas to draw on helps me interpret the problem I'm working on. Like here.

Comment (P-on-D) [reply 2019/d01p1FuelCounterUpper.d]: I submit that any programmer can read this and understand what is going on. I further submit that this code _as a solution to the AoC problem_ can be easily explained to a non-programmer.

This line alone says so much! Everything enclosed in the braces is intended to prove something meaningful. This has to be my favourite feature of D so far. I like to "sketch" in code so having a code-based canvas to draw on helps me interpret the problem I'm working on. Like here.

File: 2019/d01p1FuelCounterUpper.d [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+  import std.algorithm;
+
+  auto input = [12, 14, 1969, 100756];
```
Comment (P-on-D) [inline 2019/d01p1FuelCounterUpper.d:4]: This line establishes `input` as a list of numbers. The `auto` type specifier says "work it out for me" but really it is just less typing than `int[]` :smile:

I've written both C and C++ oldskool as well as lots of modern scripts so this syntax is both familiar and refreshing.

File: 2019/d01p1FuelCounterUpper.d [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+  auto fuelTotal = input.map!"a / 3 - 2".sum;
```
Comment (P-on-D) [inline 2019/d01p1FuelCounterUpper.d:5]: This line is where D shows a hint of its magic, specifically Uniform Function Calling Syntax (pronounced "dot") and that you can safely mix strings of code with code.

`map` is a template function that applies a function over a range of values. Here the mass-to-fuel function is provided as a string. This is not evaluated at runtime - the compiler is compiling this, with all the safeguards thereof. `a` is the name of the function's argument representing the input value.

`sum` is a function which takes an input range and returns the sum of all its values.

So, the total amount of fuel is the sum of the mass-to-fuel function applied to each of the input module mass values.

File: 2019/d01p1FuelCounterUpper.d [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+  assert(fuelTotal == 2 + 2 + 654 + 33583);
```
Comment (P-on-D) [inline 2019/d01p1FuelCounterUpper.d]: And here I prove that.

File: 2019/d01p1FuelCounterUpper.d [hunk 5]
```diff
@@ -0,0 +1,7 @@
+unittest {
+  import std.algorithm;
```
Comment (P-on-D-Reviewer) [inline 2019/d01p1FuelCounterUpper.d]: Are we supposed to ignore this line?

Comment (P-on-D) [reply 2019/d01p1FuelCounterUpper.d]: `map` and `sum` are algorithms, where else would one get them from but the standard algorithms library?

File: 2019/d01p1FuelCounterUpper.d (hunk 4 above)
Comment (P-on-D-Reviewer) [reply 2019/d01p1FuelCounterUpper.d]: How can anyone else do so?

Comment (P-on-D) [reply 2019/d01p1FuelCounterUpper.d]: Assuming my custom `d` DMD Docker image is built, `docker run --rm -v$PWD:/src d sh -c 'rdmd -unittest -main d01p1FuelCounterUpper.d`.

It does nothing.

Comment (P-on-D-Reviewer) [reply 2019/d01p1FuelCounterUpper.d]: That's worth putting in the commit message.

File: 2019/d01p1FuelCounterUpper.d [hunk 6]
```diff
@@ -1,7 +1,14 @@
...
+unittest {
   auto input = [12, 14, 1969, 100756];
   auto fuelTotal = input.map!"a / 3 - 2".sum;
-  assert(fuelTotal == 2 + 2 + 654 + 33583);
+  assert(fuelTotal == 2 + 2 + 654 + 33583); // 34241
+} version(unittest) {} else {
+
+void main() {
+  import std.conv, std.stdio;
+  stdin.byLineCopy.map!(to!int).map!"a / 3 - 2".sum.writeln;
```
Comment (P-on-D) [inline 2019/d01p1FuelCounterUpper.d:11]: Standard input is read by line, each line is converted to an integer, has the mass-to-fuel function applied, the sum is taken and printed.

I don't know yet how to explain template functions except to say that the `!` operator reveals them. New here is `to` from the standard conversions library, used to convert a string to an integer. I don't know if `byLineCopy` is needed here - there exists also `byLine` which sounds faster.

The interpretation is clear though, given the preceding unit test proving that the algorithm is correct.

Comment (P-on-D-Reviewer) [reply 2019/d01p1FuelCounterUpper.d:11]: And as seen in the third commit, it just worked with the puzzle input. No fumbling required 😄

**Pull request merged**